File: src/main/kotlin/com/example/demo/box/SnakeGame.java
//package com.example.demo.box;
//
//import com.sun.javafx.scene.traversal.Direction;
//import javafx.animation.AnimationTimer;
//import javafx.application.Application;
//import javafx.scene.Scene;
//import javafx.scene.canvas.Canvas;
//import javafx.scene.canvas.GraphicsContext;
//import javafx.scene.input.KeyCode;
//import javafx.scene.input.KeyEvent;
//import javafx.scene.layout.VBox;
//import javafx.scene.paint.Color;
//import javafx.scene.text.Font;
//import javafx.stage.Stage;
//import tornadofx.ViewTransition;
//
//import java.awt.*;
//import java.util.ArrayList;
//import java.util.List;
//import java.util.Random;
//
//public class SnakeGame extends Application {
//    static int speed = 5;
//    static int foodcolor = 0;
//    static int width = 20;
//    static int height = 20;
//    static int foodX = 0;
//    static int foodY = 0;
//    static int cornersize = 25;
//    static List<Corner> snake = new ArrayList<>();
//    static Dir direction = Dir.LEFT;
//    static boolean gameOver = false;
//    static Random rond = new Random();
//
//
//    public enum Dir {
//        LEFT, RIGHT, UP, DOWN
//    }
//
//    public static class Corner {
//        int x;
//        int y;
//
//        public Corner(int x, int y) {
//            this.x = x;
//            this.y = y;
//        }
//    }
//
//    public void start(Stage stage) {
//        try {
//            newfood();
//
//            VBox root = new VBox();
//            Canvas c = new Canvas(width * cornersize, height * cornersize);
//            GraphicsContext gc = c.getGraphicsContext2D();
//            root.getChildren().add(c);
//
//            new AnimationTimer() {
//                long lastTick = 0;
//
//                public void handle(long now) {
//                    if (lastTick == 0) {
//                        lastTick = now;
//                        tick(gc);
//                        return;
//                    }
//                    if (now - lastTick > 100000 / speed) {
//                        lastTick = now;
//                        tick(gc);
//                    }
//                }
//            }.start();
//            Scene scene = new Scene(root, width - cornersize, height * cornersize);
//
//            // control
//            scene.addEventFilter(KeyEvent.KEY_PRESSED, key -> {
//                if (key.getCode()== KeyCode.UP){ direction = Dir.UP; }
//                if (key.getCode()== KeyCode.DOWN){ direction = Dir.DOWN; }
//                if (key.getCode()== KeyCode.LEFT){ direction = Dir.LEFT; }
//                if (key.getCode()== KeyCode.RIGHT){ direction = Dir.RIGHT; }
//            });
//
//            // add start snake part
//            snake.add(new Corner(width/2,height/2));
//            snake.add(new Corner(width/2,height/2));
//            snake.add(new Corner(width/2,height/2));
//
////            scene.getStylesheets().add(getClass().getResource("application.css").toExternalForm());
//            stage.setScene(scene);
//            stage.setTitle("Snake Game");
//            stage.show();
//        } catch (Exception e){
//            e.printStackTrace();
//        }
//    }
//
//    // tick
//    public static void tick(GraphicsContext gc){
//        if (gameOver){
//            gc.setFill(Color.RED);
//            gc.setFont(new Font("",50));
//            gc.fillText("Game Over",100,250);
//            return;
//        }
//        for (int i = snake.size() - 1; i>=1;i--){
//            snake.get(i).x = snake.get(i-1).x;
//            snake.get(i).y = snake.get(i-1).y;
//        }
//        switch (direction){
//            case UP:
//                snake.get(0).y--;
//                if (snake.get(0).y<0){
//                    gameOver = true;
//                }
//            case DOWN:
//                snake.get(0).y++;
//                if (snake.get(0).y>height){
//                    gameOver = true;
//                }
//                break;
//            case LEFT:
//                snake.get(0).x--;
//                if (snake.get(0).x<0){
//                    gameOver = true;
//                }
//            case RIGHT:
//                snake.get(0).x++;
//                if (snake.get(0).x>width){
//                    gameOver = true;
//                }
//                break;
//        }
//        // eat food
//        if (foodX == snake.get(0).x && foodY == snake.get(0).y){
//                snake.add(new Corner(-1,-1));
//                newfood();
//            }
//            // self destroy
//            for (int i =1;i<snake.size();i++){
//                if (snake.get(0).x == snake.get(i).x && snake.get(0).y == snake.get(i).y) {
//                    gameOver = true;
//                    break;
//                }
//            }
//        // fill
//        // background
//        gc.setFill(Color.BLACK);
//        gc.fillRect(0,0,width*cornersize,height*cornersize);
//
//        // score
//        gc.setFill(Color.WHITE);
//        gc.setFont(new Font("",30));
//        gc.fillText("Score",+(speed-6),10,30);
//
//        //random color
//        Color cc = Color.WHITE;
//        if (foodcolor == 0) {
//            cc = Color.ANTIQUEWHITE;
//        }
//        gc.setFill(cc);
//        gc.fillOval(foodX*cornersize,foodY*cornersize,cornersize,cornersize);
//
//        // snake
//        for (Corner c:snake){
//            gc.setFill(Color.LIGHTGREEN);
//            gc.fillRect(c.x*cornersize,c.y*cornersize,cornersize-1,cornersize-1);
//            gc.setFill(Color.GREEN);
//            gc.fillRect(c.x*cornersize,c.y*cornersize,cornersize-2,cornersize-2);
//        }
//    }
//    // food
//    public static void newfood(){
//        start:while(true){
//        foodX = rond.nextInt(width);
//        foodY = rond.nextInt(height);
//
//        for (Corner c:snake){
//            if (c.x == foodX && c.y == foodY){
//                continue  start;
//            }
//        }
//        foodcolor = rond.nextInt(5);
//        speed++;
//        break;
//        }
//    }
//    public static void main(String[] args){
//        launch(args);
//    }
//}
//

File: src/main/kotlin/com/example/demo/box/SnakeGame.kt
package com.example.demo.box

import javafx.stage.Stage
import javafx.scene.layout.VBox
import javafx.scene.canvas.GraphicsContext
import javafx.animation.AnimationTimer
import javafx.application.Application
import javafx.scene.Scene
import javafx.scene.canvas.Canvas
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.paint.Color
import javafx.scene.text.Font
import tornadofx.App
import tornadofx.launch
import java.lang.Exception
import java.util.*
import kotlin.jvm.JvmStatic

fun main() {
    launch<SnakeGame>()
}
class SnakeGame : App() {
    enum class Dir {
        LEFT, RIGHT, UP, DOWN
    }

    class Corner(var x: Int, var y: Int)

    override fun start(stage: Stage) {
        try {
            newfood()
            val root = VBox()
            val c = Canvas((width * cornersize).toDouble(), (height * cornersize).toDouble())
            val gc = c.graphicsContext2D
            root.children.add(c)
            object : AnimationTimer() {
                var lastTick: Long = 0
                override fun handle(now: Long) {
                    if (lastTick == 0L) {
                        lastTick = now
                        tick(gc)
                        return
                    }
                    if (now - lastTick > 100000 / speed) {
                        lastTick = now
                        tick(gc)
                    }
                }
            }.start()
            val scene = Scene(root, (width - cornersize).toDouble(), (height * cornersize).toDouble())

            // control
            scene.addEventFilter(KeyEvent.KEY_PRESSED) { key: KeyEvent ->
                if (key.code == KeyCode.UP) {
                    direction = Dir.UP
                }
                if (key.code == KeyCode.DOWN) {
                    direction = Dir.DOWN
                }
                if (key.code == KeyCode.LEFT) {
                    direction = Dir.LEFT
                }
                if (key.code == KeyCode.RIGHT) {
                    direction = Dir.RIGHT
                }
            }

            // add start snake part
            snake.add(Corner(width / 2, height / 2))
            snake.add(Corner(width / 2, height / 2))
            snake.add(Corner(width / 2, height / 2))
//            scene.stylesheets.add(javaClass.getResource("application.css").toExternalForm())
            stage.scene = scene
            stage.title = "Snake Game"
            stage.show()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    companion object {
        var speed = 5
        var foodcolor = 0
        var width = 20
        var height = 20
        var foodX = 0
        var foodY = 0
        var cornersize = 25
        var snake: MutableList<Corner> = ArrayList()
        var direction = Dir.LEFT
        var gameOver = false
        var rond = Random()

        // tick
        fun tick(gc: GraphicsContext) {
            if (gameOver) {
                gc.fill = Color.RED
                gc.font = Font("", 50.0)
                gc.fillText("Game Over", 100.0, 250.0)
                return
            }
            for (i in snake.size - 1 downTo 1) {
                snake[i].x = snake[i - 1].x
                snake[i].y = snake[i - 1].y
            }
            when (direction) {
                Dir.UP -> {
                    snake[0].y--
                    if (snake[0].y < 0) {
                        gameOver = true
                    }
                    snake[0].y++
                    if (snake[0].y > height) {
                        gameOver = true
                    }
                }
                Dir.DOWN -> {
                    snake[0].y++
                    if (snake[0].y > height) {
                        gameOver = true
                    }
                }
                Dir.LEFT -> {
                    snake[0].x--
                    if (snake[0].x < 0) {
                        gameOver = true
                    }
                    snake[0].x++
                    if (snake[0].x > width) {
                        gameOver = true
                    }
                }
                Dir.RIGHT -> {
                    snake[0].x++
                    if (snake[0].x > width) {
                        gameOver = true
                    }
                }
            }
            // eat food
            if (foodX == snake[0].x && foodY == snake[0].y) {
                snake.add(Corner(-1, -1))
                newfood()
            }
            // self destroy
            for (i in 1 until snake.size) {
                if (snake[0].x == snake[i].x && snake[0].y == snake[i].y) {
                    gameOver = true
                    break
                }
            }
            // fill
            // background
            gc.fill = Color.BLACK
            gc.fillRect(0.0, 0.0, (width * cornersize).toDouble(), (height * cornersize).toDouble())

            // score
            gc.fill = Color.WHITE
            gc.font = Font("", 30.0)
            gc.fillText("Score", +(speed - 6).toDouble(), 10.0, 30.0)

            //random color
            var cc = Color.WHITE
            if (foodcolor == 0) {
                cc = Color.ANTIQUEWHITE
            }
            gc.fill = cc
            gc.fillOval(
                (foodX * cornersize).toDouble(),
                (foodY * cornersize).toDouble(),
                cornersize.toDouble(),
                cornersize.toDouble()
            )

            // snake
            for (c in snake) {
                gc.fill = Color.LIGHTGREEN
                gc.fillRect(
                    (c.x * cornersize).toDouble(),
                    (c.y * cornersize).toDouble(),
                    (cornersize - 1).toDouble(),
                    (cornersize - 1).toDouble()
                )
                gc.fill = Color.GREEN
                gc.fillRect(
                    (c.x * cornersize).toDouble(),
                    (c.y * cornersize).toDouble(),
                    (cornersize - 2).toDouble(),
                    (cornersize - 2).toDouble()
                )
            }
        }

        // food
        fun newfood() {
            start@ while (true) {
                foodX = rond.nextInt(width)
                foodY = rond.nextInt(height)
                for (c in snake) {
                    if (c.x == foodX && c.y == foodY) {
                        continue@start
                    }
                }
                foodcolor = rond.nextInt(5)
                speed++
                break
            }
        }
    }
}